Re-enable late joins

nfrisby opened the pull request:
Closes #1283.

[Edit: I minimized the PR and it now only does the first "primary change".]

This PR makes three primary changes.

  * It adds an idealistic "reference simulator" for `RealPBFT` and relies on that to only generate node join plans for `RealPBFT` that ensure there will always be a single shared chain and it will have at least `k` blocks in each `2k`-slot window. Node join plans that could never satisfy that chain density property were the reason late joins were recently disabled.
  * That simulator requires eg `PBftParams`. So we add a field to `TestConfig` that has a type that depends on the protocol. This "dependent generation" avoids requiring eg `Data.Reflection.Given` for the `Arbitrary (TestConfig BlockRealPBFT)` instance.
  * It consolidates the test configuration for all of the protocols we test into the same module `*.TestConfig`. I initially did this just to avoid orphan `TestableBlock` instances, but I'm wondering if we should accept such orphans in this test code and un-consolidate the `TestableBlock` instances back to the protocol-specific modules. For the few protocols we currently have, this one module seems more convenient.

The `RealPBFT` reference simulator assumes "nominal behavior", which excludes network outages, node restarts, etc. PRs adding those will either have to extend the reference simulator or else introduce a separate property that does not assume chain density etc. We'll cross that bridge when we come to it.

Looking ahead, I think always having a generator and `Property` pair for the simple case (eg there is only ever one best chain in the network) will be helpful.

Having a protocol-dependent field in the `TestConfig` will also let us extend the generators to eg also vary the protocol parameters, such as `k`, the PBFT threshold, etc.

mrBliss commented on ouroboros-consensus/test-consensus/Test/Dynamic/Util/TestConfig.hs:
```diff
@@ -0,0 +1,482 @@
+    type AdhocTestConfig BlockMockPraos = PraosParams
+
+    genTestConfig = do
+        let params   = defaultPraosParams
```
Maybe `SecurityParam 5` instead of `defaultPraosParams` is more explicit? Or do you relying `defaultPraosParams` elsewhere?

nfrisby replied:
`PraosParams` is a record type, with more than just `SecurityParam`.

nfrisby replied:
I minimized this PR, so this content will be in a different PR. Resolving this conversation.

mrBliss commented on ouroboros-consensus/test-consensus/Test/Dynamic/Util/TestConfig.hs:
```diff
@@ -0,0 +1,482 @@
+        let params   = defaultPraosParams
+            numSlots = threePraosEpochs params
+
+        defaultGenTestConfig params (NumCoreNodes 3) numSlots
```
What do you think about increasing the number of core nodes? I'm guessing it will only help find us concurrency bugs (because there will be more concurrency in each node) and problems arising from network partitionings (large/more cliques)?

nfrisby replied:
I agree. I inherited a fixed value of 3 for the `LeaderSchedule` test, and I haven't yet reconsidered it. I hope to rewrite its schedule generator soon, and that will be a good time to revisit this question.

mrBliss commented on ouroboros-consensus/test-consensus/Test/Dynamic/Util/TestConfig.hs:
```diff
@@ -0,0 +1,482 @@
+
+    go ::
+         NodeJoinPlan
+         -- ^ an /incomplete/ and /viable/ node join plain
```
```suggestion
         -- ^ an /incomplete/ and /viable/ node join plan
```

nfrisby replied:
Fixed, thanks.

mrBliss commented on ouroboros-consensus/test-consensus/Test/Dynamic/Ref/RealPBFT.hs:
```diff
@@ -0,0 +1,239 @@
+  = Absent
+    -- ^ the leader hadn't yet joined
+
+  | Nominal
```
Why `Nominal`?

nfrisby replied:
https://en.wiktionary.org/wiki/nominal#Adjective

> 10. (engineering) According to plan or design; normal.
> 
>     We'll just do a nominal flight check.
>     Apart from the slightly high temperature, all the readings from the spacecraft are nominal.

Hmm. Maybe that usage is less common than I thought. I wanted the constructor name to evoke "the boring case where the node forged a valid block that the rest of the net promptly adopted".

dcoutts replied:
Rocket scientists use "nominal". Rocket science is cool. Therefore...  :grin: 

We just want to a void a RUD in our node.

https://en.wikipedia.org/wiki/Rapid_unscheduled_disassembly

mrBliss replied:
> https://en.wiktionary.org/wiki/nominal#Adjective
> 
> > 1. (engineering) According to plan or design; normal.
> >    We'll just do a nominal flight check.
> >    Apart from the slightly high temperature, all the readings from the spacecraft are nominal.
> 
> Hmm. Maybe that usage is less common than I thought. I wanted the constructor name to evoke "the boring case where the node forged a valid block that the rest of the net promptly adopted".

Sounds totally correct, but as a non-native speaker I was not familiar with that meaning of the word. If you include a tiny comment, e.g., something like "according to plan", it's fine for me.

nfrisby replied:
I added a comment.

mrBliss commented on ouroboros-consensus/test-consensus/Test/Dynamic/Ref/RealPBFT.hs:
```diff
@@ -0,0 +1,239 @@
+-- window is the outcomes of the last @2k@ slots. We also incrementally
+-- maintain the count of 'Nominal' outcomes in the window.
+--
+data State = State !(Seq CoreNodeId) !(Seq Outcome) !NumNominals !SlotNo
```
Why not use a record?

nfrisby replied:
Fixed, thanks.

nfrisby replied:
(Just for the record, I hesitated to make it a record since: the field types were unique, I preferred a particular field order, and alphabetically field names wouldn't have preserved that order. But it was a weak preference.)

mrBliss commented on ouroboros-consensus/test-consensus/Test/Dynamic/Ref/RealPBFT.hs:
```diff
@@ -0,0 +1,239 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- | A reference simulator of the RealPBFT protocol under \"ideal
+-- circumstances\"
```
I like this idea :+1:

nfrisby replied:
Great, because I'm planning on doubling down on it! :D

I think we should have two primary test schemas: one where we generate as-rich-as-possible-yet-boring `TestConfig`s where we can predict the entire net's outcome (not sure how to handle multi-leader slots here), and one where we generate unrestrained `TestConfig`s and merely check the behavior of each node and each edge in isolation.

dcoutts replied:
That's a very similar strategy to what I'm following elsewhere for the p2p control governor. I agree it makes a lot of sense.

mrBliss commented on ouroboros-consensus/test-consensus/Test/Dynamic/Util/TestConfig.hs:
```diff
@@ -0,0 +1,482 @@
+                    (NodeJoinPlan (Map.insert nid s' m))
+                    st
+        s' <- genSlot (Ref.RealPBFT.nextSlot st) lastSlot
+            `suchThat` check
```
I have recently been bitten by `suchThat`: in some rare cases, it was impossible to generate values that satisfied the condition, resulting in hangs in my testsuite. I only noticed it after running > 10k tests. Debugging it was very annoying, mostly because I excluded the possibility that my use of `suchThat` was the cause, as I couldn't conceive of the rare case that triggered the infinite loop.
</war story>

Can you convince yourself that this will never result in an infinite loop? Moreover, will it not needlessly make the testsuite slower (the RealPBFT protocol tests are already slow enough, you probably suffer the most from this already)?

nfrisby replied:
Yes, I have suffered by the same thing while working on this test suite! `suchThat` is certainly a double-edged sword. It's very high on my list of suspects for "why is this test hanging?"

So I've been careful in this use of itr, yes. I'll try to elaborate in the comments. I've also been considering a `suchThatNTimes` combinator derived from `Quickcheck`'s `suchThatMaybe`.

nfrisby replied:
I realized the `check` predicate is monotonic. So we can use a binary search instead of `suchThat`: see the `test-consensus: replace suchThat with a binary search` commit.

mrBliss commented on ouroboros-consensus/test-consensus/Test/Dynamic/Util/TestConfig.hs:
```diff
@@ -0,0 +1,482 @@
+{-# LANGUAGE StandaloneDeriving  #-}
+{-# LANGUAGE TypeFamilies        #-}
+
+-- | Test configuration, including per-protocol config
```
I'm fine with splitting this up in a module per protocol, even if that requires orphan instances. Up to you.

nfrisby replied:
I minimized this PR, so this content will be in a different PR. Resolving this conversation.

nfrisby commented on ouroboros-consensus/test-consensus/Test/Dynamic/Util/TestConfig.hs:
```diff
@@ -0,0 +1,482 @@
+    type AdhocTestConfig BlockMockBFT = SecurityParam
+
+    genTestConfig = join $
+      defaultGenTestConfig (SecurityParam 5)
```
It's tough to say. I would guess "Yes, but it might just be an uninteresting hiccup".

I know that protocols with anachrony limits (just `RealPBFT` right now) may fail in various ways depending on the `k/N/t` ratio (where `t` is the PBFT signature threshold). But in production we always have `k >> N`, right?

We also use a relatively small number of slots `S` in each test run (`S <= 100`, right now); maybe large `S/k` ratios might challenge some assumptions.

nfrisby replied:
I minimized this PR, so this content will be in a different PR. Resolving this conversation.

nfrisby commented on ouroboros-consensus/test-consensus/Test/Dynamic/Util/TestConfig.hs:
```diff
@@ -0,0 +1,482 @@
+-- | Protocol-dependent configuration
+--
+class Show (AdhocTestConfig blk) => TestableBlock blk where
+    type AdhocTestConfig blk
```
Ah yes, the associated type itself deserves a comment.

> I am surprised that it is not used by any of the methods of this type class.

It is but only indirectly via `TestConfig`.

nfrisby replied:
I minimized this PR, so this content will be in a different PR. Resolving this conversation.

nfrisby commented:
OK. I just did a force push that drastically minimizes this PR. It's no longer trying to share much functionality -- I had gotten ahead of myself with that. I replaced it all with just using a `forAllShrink` in `RealPBFT.hs`.

I am tempted to do the reorg/consolidation, but it need not be in this PR.

nfrisby commented on ouroboros-consensus/test-consensus/Test/Dynamic/Util/TestConfig.hs:
```diff
@@ -0,0 +1,482 @@
+    -- | List any inconsistencies between 'adhocTestConfig' and the other
+    -- fields
+    --
+    -- EG 'adhocTestConfig' might include on 'NumCoreNodes'
```
I minimized this PR, so this content will be in a different PR. Resolving this conversation.

nfrisby commented on ouroboros-consensus/test-consensus/Test/Dynamic/Util/TestConfig.hs:
```diff
@@ -0,0 +1,482 @@
+
+-- | Shrink, including the number of slots but not number of nodes
+--
+shrinkTestConfigSlotsOnly :: TestConfig blk-> [TestConfig blk]
```
Fixed, thanks.

nfrisby commented on ouroboros-consensus/test-consensus/Test/Dynamic/Util/TestConfig.hs:
```diff
@@ -0,0 +1,482 @@
+
+-- | Shrink, including the number of nodes and slots
+--
+shrinkTestConfigFreely :: TestConfig blk-> [TestConfig blk]
```
I minimized this PR, so this content will be in a different PR. Resolving this conversation.

nfrisby commented:
Rebased again for typos and a new file I forgot to `git add`.

nfrisby commented on ouroboros-consensus/test-consensus/Test/Dynamic/RealPBFT.hs:
```diff
@@ -177,4 +198,115 @@ generateGenesisConfig (NumCoreNodes n) =
               -- The nodes are the richmen
               { Genesis.tboRichmen = fromIntegral n }
         }
+      , Genesis.gsK = Common.BlockCount $ maxRollbacks pbftSecurityParam
```
This was 10 before, but now it's 1) easier to see that and 2) easy to vary it if we want.

nfrisby commented:
Two things just happened: I rebased onto master and then learned that the tests now fail :(

Before I noticed that, I replaced `suchThat` with a binary search. So I rebased back to where there tests were passing and also added the commit that drops `suchThat`.

I'm now locally rebasing to the latest master and debugging the new test failure.

nfrisby commented:
Latest rebase does two things:

  1. Rebases forward to the latest master that does not fail the test -- i.e. the one prior to #1307
  1. Simplifies the binary search.

mrBliss commented on ouroboros-consensus/test-consensus/Test/Dynamic/Ref/RealPBFT.hs:
```diff
@@ -0,0 +1,255 @@
+-- | The state of the RealPBFT net
+--
+data State = State
+  { ids      :: !(Seq CoreNodeId)
```
Maybe `forgers` would be a more self-explanatory name?

nfrisby replied:
Fixed, thanks.

mrBliss commented on ouroboros-consensus/test-consensus/Test/Dynamic/Ref/RealPBFT.hs:
```diff
@@ -0,0 +1,255 @@
+--
+extendId :: PBftParams -> State -> CoreNodeId -> State
+extendId params st@State{ids} i = st
+  { ids      = snd $ prune (oneK params) $ ids Seq.|> i
```
```suggestion
  { ids = snd $ prune (oneK params) $ ids Seq.|> i
```

nfrisby replied:
Fixed, thanks.

nfrisby commented on ouroboros-consensus/test-consensus/Test/Dynamic/Util/TestConfig.hs:
```diff
@@ -0,0 +1,482 @@
+-- | Shrink, including the number of slots but not number of nodes
+--
+shrinkTestConfigSlotsOnly :: TestConfig blk-> [TestConfig blk]
+shrinkTestConfigSlotsOnly TestConfig
```
I simplified it so that we have two shrinks, currently:

  * One in the `Arbitrary TestConfig` instance that shrinks all fields.
  * One in `RealPBFT.hs` that shrinks everything except num core nodes, since doing that might violate the chain density invariant that the custom `NodeJoinPlan` generator in that module ensures.

`Praos.hs` and `LeaderSchedule.hs` use nested `forAllShrink`s, since they have some fixed parameters (which have always been fixed and I haven't yet considered varying).

nfrisby commented:
I rebased to

  1) Include `origin/master`
  2) Squash commits via `fixup`

nfrisby commented:
bors r+
